packages/block-editor/src/components/inner-blocks/use-inner-block-template-sync.js

```diff
@@ -44,25 +44,31 @@ export default function useInnerBlockTemplateSync(
 		useSelect( blockEditorStore );
 	const { replaceInnerBlocks, __unstableMarkNextChangeAsNotPersistent } =
 		useDispatch( blockEditorStore );
-	const innerBlocks = useSelect(
-		( select ) => select( blockEditorStore ).getBlocks( clientId ),
+	const { innerBlocks } = useSelect(
+		( select ) => ( {
+			innerBlocks: select( blockEditorStore ).getBlocks( clientId ),
+		} ),
 		[ clientId ]
 	);
-	const { getBlocks } = useSelect( blockEditorStore );
 
 	// Maintain a reference to the previous value so we can do a deep equality check.
 	const existingTemplate = useRef( null );
 	useLayoutEffect( () => {
+		let isCancelled = false;
+
 		// There's an implicit dependency between useInnerBlockTemplateSync and useNestedSettingsUpdate
 		// The former needs to happen after the latter and since the latter is using microtasks to batch updates (performance optimization),
 		// we need to schedule this one in a microtask as well.
-		// Exemple: If you remove queueMicrotask here, ctrl + click to insert quote block won't close the inserter.
+		// Example: If you remove queueMicrotask here, ctrl + click to insert quote block won't close the inserter.
 		window.queueMicrotask( () => {
+			if ( isCancelled ) {
+				return;
+			}
+
 			// Only synchronize innerBlocks with template if innerBlocks are empty
 			// or a locking "all" or "contentOnly" exists directly on the block.
-			const currentInnerBlocks = getBlocks( clientId );
```

Comment:
I'm pretty sure there was a bug that forced me to use this selector here. In other words, something updated the inner blocks between the scheduling of the micro task and its execution. I can't recall exactly though and looking at the PR that introduced the microtask, it's not clear either.

Reply:
> This approach also allows us to avoid reading currentInnerBlocks. Cancelling provides a guarantee that the innerBlocks from the closure are always valid.

It allows this only if the "effect cleanup" is triggered before the microtask. Is this a given?

Reply:
> Is this a given?

Thanks for raising this. I investigated on a little toy project and the answer is that with a correct implementation of `useSelect`, yes, it should be a given. When the canonical store value changes, a store listener is called, `useSelect` calls `forceRender()` (which is an alias for `setState`), and that `setState` performs effect cleanup. If all of that happens synchronously within one event loop tick, then there is no room for another microtask to change the value.

Whether it happens synchronously depends on the exact setup of the environment. React 18 with legacy `render` (React 17) is sync, React 18 with `createRoot` is not. We'd have to use `useSyncExternalStore` in `useSelect` to make it sync again.

I'll try to rewrite the implementation to be more resilient and not require complex reasoning about React scheduling to prove that it's correct.

Reply:
I did some experiments also with `useSyncExternalStore`, `use[Layout]Effect` and `queueMicrotask`, and the conclusion is that `innerBlocks` and `currentInnerBlocks` can get out of sync quite easily. So I reverted the part of my patch where I removed `currentInnerBlocks`.

Now the patch just cancels the scheduled microtask every time the effect is cleaned up. Either on unmount, or when relevant data (inner blocks or template lock) change.